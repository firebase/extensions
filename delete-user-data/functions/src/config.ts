/*
 * Copyright 2019 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export default {
  location: process.env.LOCATION,
  firestorePaths: process.env.FIRESTORE_PATHS,
  firestoreDeleteMode: process.env.FIRESTORE_DELETE_MODE,
  rtdbPaths: process.env.RTDB_PATHS,
  storagePaths: process.env.STORAGE_PATHS,
  storageBucketDefault: process.env.STORAGE_BUCKET,
  SELECTED_DATABASE_INSTANCE: process.env.SELECTED_DATABASE_INSTANCE,
};
