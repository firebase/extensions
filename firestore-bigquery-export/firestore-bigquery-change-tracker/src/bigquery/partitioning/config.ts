export type TimePartitioningGranularity = "HOUR" | "DAY" | "MONTH" | "YEAR";

export type PartitioningFieldType = "TIMESTAMP" | "DATE" | "DATETIME";

export interface BasePartitioningConfig {
  granularity?: TimePartitioningGranularity | "NONE" | null;
  bigqueryColumnName?: string;
  bigqueryColumnType?: PartitioningFieldType;
  firestoreFieldName?: string;
}

export interface NoPartitioning extends BasePartitioningConfig {
  granularity?: "NONE" | null | undefined;
  bigqueryColumnName?: undefined;
  bigqueryColumnType?: undefined;
  firestoreFieldName?: undefined;
}

export interface IngestionTimePartitioning extends BasePartitioningConfig {
  granularity: TimePartitioningGranularity;
  bigqueryColumnName?: undefined;
  bigqueryColumnType?: undefined;
  firestoreFieldName?: undefined;
}

export interface FirestoreTimestampPartitioning extends BasePartitioningConfig {
  granularity: TimePartitioningGranularity;
  bigqueryColumnName: "timestamp";
  bigqueryColumnType?: PartitioningFieldType;
  firestoreFieldName?: undefined;
}

export interface FirestoreFieldPartitioning extends BasePartitioningConfig {
  granularity: TimePartitioningGranularity;
  bigqueryColumnName: string;
  bigqueryColumnType: PartitioningFieldType;
  firestoreFieldName: string;
}

export type PartitioningStrategy =
  | NoPartitioning
  | IngestionTimePartitioning
  | FirestoreTimestampPartitioning
  | FirestoreFieldPartitioning;

export enum PartitioningType {
  NONE = "NONE",
  INGESTION_TIME = "INGESTION_TIME",
  FIRESTORE_TIMESTAMP = "FIRESTORE_TIMESTAMP",
  FIRESTORE_FIELD = "FIRESTORE_FIELD",
}

export class PartitioningConfig {
  private strategy: PartitioningStrategy;
  private type: PartitioningType;

  constructor(config: PartitioningStrategy) {
    this.strategy = config;
    this.type = this.determineType(config);
  }

  private determineType(config: PartitioningStrategy): PartitioningType {
    if (!config.granularity || config.granularity === "NONE") {
      return PartitioningType.NONE;
    }

    if (!config.bigqueryColumnName && !config.firestoreFieldName) {
      return PartitioningType.INGESTION_TIME;
    }

    if (
      config.bigqueryColumnName === "timestamp" &&
      !config.firestoreFieldName
    ) {
      return PartitioningType.FIRESTORE_TIMESTAMP;
    }

    if (
      config.bigqueryColumnName &&
      config.bigqueryColumnType &&
      config.firestoreFieldName
    ) {
      return PartitioningType.FIRESTORE_FIELD;
    }

    return PartitioningType.NONE;
  }

  getType(): PartitioningType {
    return this.type;
  }

  getStrategy(): PartitioningStrategy {
    return this.strategy;
  }

  isNoPartitioning(): boolean {
    return this.type === PartitioningType.NONE;
  }

  isIngestionTimePartitioning(): boolean {
    return this.type === PartitioningType.INGESTION_TIME;
  }

  isFirestoreTimestampPartitioning(): boolean {
    return this.type === PartitioningType.FIRESTORE_TIMESTAMP;
  }

  isFirestoreFieldPartitioning(): boolean {
    return this.type === PartitioningType.FIRESTORE_FIELD;
  }

  getGranularity(): TimePartitioningGranularity | "NONE" | null | undefined {
    return this.strategy.granularity;
  }

  getBigQueryColumnName(): string | undefined {
    return this.strategy.bigqueryColumnName;
  }

  getBigQueryColumnType(): PartitioningFieldType | undefined {
    return this.strategy.bigqueryColumnType;
  }

  getFirestoreFieldName(): string | undefined {
    return this.strategy.firestoreFieldName;
  }

  static none(): PartitioningConfig {
    return new PartitioningConfig({ granularity: "NONE" });
  }

  static ingestionTime(
    granularity: TimePartitioningGranularity
  ): PartitioningConfig {
    return new PartitioningConfig({ granularity });
  }

  static firestoreTimestamp(
    granularity: TimePartitioningGranularity,
    columnType?: PartitioningFieldType
  ): PartitioningConfig {
    return new PartitioningConfig({
      granularity,
      bigqueryColumnName: "timestamp",
      bigqueryColumnType: columnType,
    });
  }

  static firestoreField(
    granularity: TimePartitioningGranularity,
    bigqueryColumnName: string,
    bigqueryColumnType: PartitioningFieldType,
    firestoreFieldName: string
  ): PartitioningConfig {
    return new PartitioningConfig({
      granularity,
      bigqueryColumnName,
      bigqueryColumnType,
      firestoreFieldName,
    });
  }
}
