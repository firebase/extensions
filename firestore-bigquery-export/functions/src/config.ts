/*
 * Copyright 2019 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

function timePartitioning(type) {
  if (
    type === "HOUR" ||
    type === "DAY" ||
    type === "MONTH" ||
    type === "YEAR"
  ) {
    return type;
  }

  return null;
}

export function clustering(clusters: string | undefined) {
  return clusters ? clusters.split(",").slice(0, 4) : null;
}

export default {
  bqProjectId: process.env.BIGQUERY_PROJECT_ID,
  collectionPath: process.env.COLLECTION_PATH,
  datasetId: process.env.DATASET_ID,
  tableId: process.env.TABLE_ID,
  location: process.env.LOCATION,
  initialized: false,
  datasetLocation: process.env.DATASET_LOCATION,
  backupCollectionId: process.env.BACKUP_COLLECTION,
  transformFunction: process.env.TRANSFORM_FUNCTION,
  timePartitioning: timePartitioning(process.env.TABLE_PARTITIONING),
  timePartitioningField: process.env.TIME_PARTITIONING_FIELD,
  timePartitioningFieldType:
    process.env.TIME_PARTITIONING_FIELD_TYPE !== "omit"
      ? process.env.TIME_PARTITIONING_FIELD_TYPE
      : undefined,
  timePartitioningFirestoreField: process.env.TIME_PARTITIONING_FIRESTORE_FIELD,
  clustering: clustering(process.env.CLUSTERING),
  wildcardIds: process.env.WILDCARD_IDS === "true" ? true : false,
};
