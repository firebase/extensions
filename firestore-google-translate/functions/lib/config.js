"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = {
    languages: process.env.LANGUAGES.split(","),
    messageFieldName: process.env.MESSAGE_FIELD_NAME,
    translationsFieldName: process.env.TRANSLATIONS_FIELD_NAME,
};
