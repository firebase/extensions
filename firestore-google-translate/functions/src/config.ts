export default {
  languages: process.env.LANGUAGES.split(","),
  location: process.env.LOCATION,
  messageFieldName: process.env.MESSAGE_FIELD_NAME,
  translationsFieldName: process.env.TRANSLATIONS_FIELD_NAME,
};
