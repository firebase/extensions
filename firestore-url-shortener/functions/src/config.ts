export default {
  bitlyAccessToken: process.env.BITLY_ACCESS_TOKEN,
  collectionPath: process.env.COLLECTION_PATH,
  location: process.env.LOCATION,
  shortUrlFieldName: process.env.SHORT_URL_FIELD_NAME,
  urlFieldName: process.env.URL_FIELD_NAME,
};
