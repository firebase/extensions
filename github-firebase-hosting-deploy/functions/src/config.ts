export default {
  branch: process.env.BRANCH,
  deployRoot: process.env.DEPLOY_ROOT,
  location: process.env.LOCATION,
  repository: process.env.GITHUB_REPOSITORY,
  siteName: process.env.SITE_NAME,
};
