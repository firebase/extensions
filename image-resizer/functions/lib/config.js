"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = {
    cacheControlHeader: process.env.CACHE_CONTROL_HEADER,
    maxHeight: process.env.IMG_MAX_HEIGHT,
    maxWidth: process.env.IMG_MAX_WIDTH,
    signedUrlsPath: process.env.SIGNED_URLS_PATH,
    signedUrlsExpirationDate: process.env.SIGNED_URLS_EXPIRATION_DATE,
};
