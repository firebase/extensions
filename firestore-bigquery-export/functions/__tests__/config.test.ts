import { readFileSync } from "fs";

import { resolve as pathResolve } from "path";

import * as yaml from "js-yaml";
import mockedEnv from "mocked-env";
import config from "../src/config";

let restoreEnv;
let extensionYaml;
let extensionParams;

const environment = {
  LOCATION: "us-central1",
  DATASET_ID: "my_dataset",
  TABLE_ID: "my_table",
};

describe("extension config", () => {
  beforeAll(() => {
    extensionYaml = yaml.safeLoad(
      readFileSync(pathResolve(__dirname, "../../extension.yaml"), "utf8")
    );

    extensionParams = extensionYaml.params.reduce((obj, param) => {
      obj[param.param] = param;
      return obj;
    }, {});
  });

  beforeEach(() => {
    restoreEnv = mockedEnv(environment);
  });

  afterEach(() => restoreEnv());

  test("config loaded from environment variables", () => {
    const functionsConfig = config;

    expect(functionsConfig).toMatchSnapshot({});
  });

  // DATASET_ID
  describe("config.datasetId", () => {
    test("param exists", () => {
      const extensionParam = extensionParams["DATASET_ID"];
      expect(extensionParam).toMatchSnapshot();
    });

    describe("validationRegex", () => {
      test("does not allow empty strings", () => {
        const { validationRegex } = extensionParams["DATASET_ID"];
        expect(Boolean("".match(new RegExp(validationRegex)))).toBeFalsy();
      });

      test("does not allow spaces", () => {
        const { validationRegex } = extensionParams["DATASET_ID"];
        expect(
          Boolean("foo bar,".match(new RegExp(validationRegex)))
        ).toBeFalsy();
      });

      test("allows a alphanumeric underscore ids", () => {
        const { validationRegex } = extensionParams["DATASET_ID"];
        expect(
          Boolean("my_dataset".match(new RegExp(validationRegex)))
        ).toBeTruthy();
      });
    });
  });

  // TABLE_ID
  describe("config.tableId", () => {
    test("param exists", () => {
      const extensionParam = extensionParams["TABLE_ID"];
      expect(extensionParam).toMatchSnapshot();
    });

    describe("validationRegex", () => {
      test("does not allow empty strings", () => {
        const { validationRegex } = extensionParams["TABLE_ID"];
        expect(Boolean("".match(new RegExp(validationRegex)))).toBeFalsy();
      });
      test("does not allow spaces", () => {
        const { validationRegex } = extensionParams["TABLE_ID"];
        expect(
          Boolean("foo bar,".match(new RegExp(validationRegex)))
        ).toBeFalsy();
      });

      test("allows a alphanumeric underscore ids", () => {
        const { validationRegex } = extensionParams["TABLE_ID"];
        expect(
          Boolean("my_table".match(new RegExp(validationRegex)))
        ).toBeTruthy();
      });
    });
  });
});
