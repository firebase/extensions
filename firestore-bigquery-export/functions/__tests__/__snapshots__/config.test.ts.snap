// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extension config config loaded from environment variables 1`] = `
Object {
  "collectionPath": undefined,
  "datasetId": "my_dataset",
  "datasetLocation": undefined,
  "initialized": false,
  "location": "us-central1",
  "tableId": "my_table",
  "tablePartitioning": null,
}
`;

exports[`extension config config.datasetId param exists 1`] = `
Object {
  "default": "firestore_export",
  "description": "What ID would you like to use for your BigQuery dataset? This extension will create the dataset, if it doesn't already exist.",
  "example": "firestore_export",
  "label": "Dataset ID",
  "param": "DATASET_ID",
  "required": true,
  "type": "string",
  "validationErrorMessage": "BigQuery dataset IDs must be alphanumeric (plus underscores) and must be no more than 1024 characters.
",
  "validationRegex": "^[a-zA-Z0-9_]+$",
}
`;

exports[`extension config config.tableId param exists 1`] = `
Object {
  "default": "posts",
  "description": "What identifying prefix would you like to use for your table and view inside your BigQuery dataset? This extension will create the table and view, if they don't already exist.",
  "example": "posts",
  "label": "Table ID",
  "param": "TABLE_ID",
  "required": true,
  "type": "string",
  "validationErrorMessage": "BigQuery table IDs must be alphanumeric (plus underscores) and must be no more than 1024 characters.
",
  "validationRegex": "^[a-zA-Z0-9_]+$",
}
`;
